Type hinting

I must say that I really start to dislike the dynamic typing nature of Python when it's about function signatures.

Type hinting as specified in [PEP484](https://www.python.org/dev/peps/pep-0484/) would be cool to solve this issue.

Somewhat related to #9

[`mypy`](https://mypy.readthedocs.io/en/stable/) seem to be a nice type checker.

Using it via `mypy --disallow-untyped-defs src/`, it can show all functions which are not statically typed yet:

```
> mypy --disallow-untyped-defs src/
src/util/split.py:4: error: Function is missing a type annotation
src/util/rot.py:4: error: Function is missing a type annotation
src/util/rot.py:16: error: Function is missing a type annotation
src/util/encode.py:7: error: Function is missing a type annotation
src/util/encode.py:12: error: Function is missing a type annotation
src/util/encode.py:17: error: Function is missing a type annotation
src/util/encode.py:20: error: Function is missing a type annotation
src/util/encode.py:26: error: Function is missing a type annotation
src/util/encode.py:29: error: Function is missing a type annotation
src/util/concat_bits.py:4: error: Function is missing a type annotation
src/util/min_bits.py:6: error: Function is missing a type annotation
src/ciphers/modi/ecb.py:16: error: Function is missing a type annotation
src/ciphers/modi/ecb.py:26: error: Function is missing a type annotation
src/ciphers/modi/ecb.py:36: error: Function is missing a type annotation
src/util/wrap.py:19: error: Function is missing a type annotation
src/util/wrap.py:130: error: Function is missing a type annotation
src/util/wrap.py:133: error: Function is missing a type annotation
src/util/wrap.py:139: error: Function is missing a type annotation
src/util/wrap.py:144: error: Function is missing a type annotation
src/util/wrap.py:147: error: Function is missing a type annotation
src/ciphers/feal.py:53: error: Function is missing a type annotation
src/ciphers/feal.py:90: error: Function is missing a type annotation
src/ciphers/feal.py:95: error: Function is missing a type annotation
src/ciphers/feal.py:100: error: Function is missing a type annotation
src/ciphers/feal.py:105: error: Function is missing a type annotation
src/ciphers/feal.py:130: error: Function is missing a type annotation
src/ciphers/feal.py:151: error: Function is missing a type annotation
src/ciphers/feal.py:158: error: Function is missing a type annotation
src/ciphers/feal.py:165: error: Function is missing a type annotation
src/ciphers/feal.py:173: error: Function is missing a type annotation
src/ciphers/feal.py:181: error: Function is missing a type annotation
src/ciphers/feal.py:205: error: Function is missing a type annotation
src/ciphers/feal.py:229: error: Function is missing a return type annotation
Found 33 errors in 8 files (checked 12 source files)
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
 venv
 __pycache__
 .coverage
+.mypy_cache

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,9 @@
 coverage==5.1
 docopt==0.6.2
+mypy==0.770
+mypy-extensions==0.4.3
 pycodestyle==2.6.0
 pydocstyle==5.0.2
 snowballstemmer==2.0.0
+typed-ast==1.4.1
+typing-extensions==3.7.4.2

--- a/src/ciphers/feal.py
+++ b/src/ciphers/feal.py
@@ -37,6 +37,7 @@ Usage:
 
 import sys
 from pathlib import Path
+from typing import Sequence, Tuple, Any, Optional
 
 from docopt import docopt  # type: ignore
 
@@ -50,7 +51,7 @@ from util.rot import rot_left
 from util.split import split
 
 
-def key_schedule(key, n=32):
+def key_schedule(key: int, n: int = 32) -> Sequence[int]:
     """Return the subkeys created by the key scheduler of FEAL-NX.
 
     Creates the N+8 16-bit subkeys which are needed during en-/decryption.
@@ -87,22 +88,22 @@ def key_schedule(key, n=32):
     return k
 
 
-def s0(a, b):
+def s0(a: int, b: int) -> int:
     """Return substitution value of S-Box 0."""
     return _s(a, b, 0)
 
 
-def s1(a, b):
+def s1(a: int, b: int) -> int:
     """Return substitution value of S-Box 1."""
     return _s(a, b, 1)
 
 
-def _s(a, b, i):
+def _s(a: int, b: int, i: int) -> int:
     """General substitution box implementation for FEAL-NX."""
     return rot_left((a + b + i) % 256, 2, 8)
 
 
-def f(a, b):
+def f(a: int, b: int) -> int:
     """f-function of FEAL-NX.
 
     a must be 32-bit and b must be 16-bit long.
@@ -114,20 +115,20 @@ def f(a, b):
         raise ValueError("a key must be 32-bit")
     if b >= 2 ** 16:
         raise ValueError("b key must be 16-bit")
-    a = split(n=4, size=8, bits=a)
-    b = split(n=2, size=8, bits=b)
-    f1 = a[1] ^ b[0]
-    f2 = a[2] ^ b[1]
-    f1 ^= a[0]
-    f2 ^= a[3]
+    a_k = split(n=4, size=8, bits=a)
+    b_k = split(n=2, size=8, bits=b)
+    f1 = a_k[1] ^ b_k[0]
+    f2 = a_k[2] ^ b_k[1]
+    f1 ^= a_k[0]
+    f2 ^= a_k[3]
     f1 = s1(f1, f2)
     f2 = s0(f2, f1)
-    f0 = s0(a[0], f1)
-    f3 = s1(a[3], f2)
+    f0 = s0(a_k[0], f1)
+    f3 = s1(a_k[3], f2)
     return concat_bits(f0, f1, f2, f3, n=8)
 
 
-def fk(a, b):
+def fk(a: int, b: int) -> int:
     """f_k-function of FEAL-NX.
 
     Input keys must be 32-bit.
@@ -137,32 +138,33 @@ def fk(a, b):
     """
     if a >= 2 ** 32 or b >= 2 ** 32:
         raise ValueError("Input keys must be 32-bit")
-    a = split(n=4, size=8, bits=a)
-    b = split(n=4, size=8, bits=b)
-    fk1 = a[1] ^ a[0]
-    fk2 = a[2] ^ a[3]
-    fk1 = s1(fk1, fk2 ^ b[0])
-    fk2 = s0(fk2, fk1 ^ b[1])
-    fk0 = s0(a[0], fk1 ^ b[2])
-    fk3 = s1(a[3], fk2 ^ b[3])
+    a_k = split(n=4, size=8, bits=a)
+    b_k = split(n=4, size=8, bits=b)
+    fk1 = a_k[1] ^ a_k[0]
+    fk2 = a_k[2] ^ a_k[3]
+    fk1 = s1(fk1, fk2 ^ b_k[0])
+    fk2 = s0(fk2, fk1 ^ b_k[1])
+    fk0 = s0(a_k[0], fk1 ^ b_k[2])
+    fk3 = s1(a_k[3], fk2 ^ b_k[3])
     return concat_bits(fk0, fk1, fk2, fk3, n=8)
 
 
-def _encrypt_preprocessing(subkeys, text):
+def _encrypt_preprocessing(subkeys: Sequence[int], text: int) -> int:
     p = text ^ concat_bits(*subkeys, n=16)
     l0, r0 = split(2, 32, p)
     p ^= l0
     return p
 
 
-def _decrypt_preprocessing(subkeys, text):
+def _decrypt_preprocessing(subkeys: Sequence[int], text: int) -> int:
     p = text ^ concat_bits(*subkeys, n=16)
     rn, ln = split(2, 32, p)
     p = concat_bits(rn, ln, n=32) ^ rn
     return p
 
 
-def _encrypt_iterative_calculation(l0, r0, sk, n=32):
+def _encrypt_iterative_calculation(l0: int, r0: int, sk: Sequence[int], n: int = 32) \
+        -> Tuple[Sequence[int], Sequence[int]]:
     l, r = [l0], [r0]
     for i in range(1, n + 1):
         r.append(l[i - 1] ^ f(r[i - 1], sk[i - 1]))
@@ -170,15 +172,16 @@ def _encrypt_iterative_calculation(l0, r0, sk, n=32):
     return l, r
 
 
-def _decrypt_iterative_calculation(ln, rn, sk, n=32):
-    l, r = [None] * n + [ln], [None] * n + [rn]
+def _decrypt_iterative_calculation(ln: int, rn: int, sk: Sequence[int], n: int = 32) \
+        -> Tuple[Sequence[int], Sequence[int]]:
+    l, r = [0] * n + [ln], [0] * n + [rn]
     for i in reversed(range(1, n + 1)):
         l[i - 1] = r[i] ^ f(l[i], sk[i - 1])
         r[i - 1] = l[i]
     return l, r
 
 
-def encrypt(key, text, **kwargs):
+def encrypt(key: int, text: int, *args: Any, **kwargs: Any) -> int:
     """Encrypt the text with the given key using FEAL-NX encryption.
 
     Raises error if text is longer than 64-bit or key is longer than 128-bit.
@@ -202,7 +205,7 @@ def encrypt(key, text, **kwargs):
     return c
 
 
-def decrypt(key, text, **kwargs):
+def decrypt(key: int, text: int, *args: Any, **kwargs: Any) -> int:
     """Decrypt the ciphertext with the given key using FEAL-NX decryption.
 
     Raises error if text is longer than 64-bit or key is longer than 128-bit.
@@ -226,7 +229,7 @@ def decrypt(key, text, **kwargs):
     return p
 
 
-def feal():
+def feal() -> Optional[int]:
     """Execute FEAL-NX cipher with arguments given on comand line.
 
     Gets arguments from docopt which parses sys.argv.
@@ -245,12 +248,12 @@ def feal():
         return _encrypt(k, text, n=n)
     elif args['decrypt']:
         return _decrypt(k, text, n=n)
+    return None
 
 
 if __name__ == "__main__":
     try:
-        f = feal()
-        print(f)
+        print(feal())
     except ValueError as e:
         print(e)
         exit(1)

--- a/src/ciphers/feal.py
+++ b/src/ciphers/feal.py
@@ -38,7 +38,7 @@ Usage:
 import sys
 from pathlib import Path
 
-from docopt import docopt
+from docopt import docopt  # type: ignore
 
 # make sure that following imports can be resolved when executing this script from cmdline
 

--- a/src/ciphers/modi/ecb.py
+++ b/src/ciphers/modi/ecb.py
@@ -7,13 +7,15 @@ The en- or decrypted blocks are then concatenated.
 """
 
 from math import ceil
+from typing import Any
 
 from util.concat_bits import concat_bits
 from util.min_bits import min_bits
 from util.split import split
+from util.types import CipherFunction
 
 
-def ecb(cipher_fn, blocksize):
+def ecb(cipher_fn: CipherFunction, blocksize: int) -> CipherFunction:
     """Electronic Code Book Mode (ECB) implementation for ciphers.
 
     Works by splitting the argument into the needed blocks, passing those blocks to the
@@ -23,7 +25,7 @@ def ecb(cipher_fn, blocksize):
     :param blocksize        Block size of cipher in bits on which we want to use ECB mode
     """
 
-    def _ecb(key, text, *args, **kwargs):
+    def _ecb(key: int, text: int, *args: Any, **kwargs: Any) -> int:
         # calculate amount of blocks
         n = ceil(min_bits(text) / blocksize)
         in_blocks = split(n, blocksize, text)
@@ -33,6 +35,6 @@ def ecb(cipher_fn, blocksize):
     return _ecb
 
 
-def feal_ecb(cipher_fn):
+def feal_ecb(cipher_fn: CipherFunction) -> CipherFunction:
     """FEAL cipher functions wrapped with ECB."""
     return ecb(cipher_fn, blocksize=64)

--- a/src/util/concat_bits.py
+++ b/src/util/concat_bits.py
@@ -1,17 +1,17 @@
 """Exports function which concatenates bitstrings into one bitstring."""
 
 
-def concat_bits(*args, n):
+def concat_bits(*args: int, n: int) -> int:
     """Concatenates the given bitstring arguments into one bitstring.
 
     The first key is shifted to the most significant n-bits, the second key to the second most significant n-bits
     and so forth.
     Raises error when one subkey is larger than n.
     """
-    args = list(args)  # convert args to list because *args creates tuple
-    args.reverse()
+    args_ = list(args)  # convert args to list because *args creates tuple
+    args_.reverse()
     r = 0
-    for i, arg in enumerate(args):
+    for i, arg in enumerate(args_):
         if arg > 2 ** n:
             raise ValueError("subkey is larger than slot given by n")
         r |= arg << (i * n)

--- a/src/util/encode.py
+++ b/src/util/encode.py
@@ -2,31 +2,32 @@
 
 Also exports encoding wrapper for ciphers.
 """
+from util.types import CipherFunction, Any
 
 
-def encode(text, encoding='utf8'):
+def encode(text: str, encoding: str = 'utf8') -> int:
     """Encode the input text with the given encoding and returns a number."""
     return int(text.encode(encoding).hex(), 16)
 
 
-def decode(number, encoding='utf8'):
+def decode(number: int, encoding: str = 'utf8') -> str:
     """Return the string which the given number represents in the given encoding."""
     return bytes.fromhex(hex(number)[2:]).decode(encoding)
 
 
-def encode_wrapper(cipher_fn):
+def encode_wrapper(cipher_fn: CipherFunction) -> CipherFunction:
     """Return wrapper for cipher functions to encode the text input before passing it to the cipher function."""
 
-    def cipher_fn_wrapper(key, text, *args, **kwargs):
+    def cipher_fn_wrapper(key: int, text: str, *args: Any, **kwargs: Any) -> int:
         return cipher_fn(key, encode(text), *args, **kwargs)
 
     return cipher_fn_wrapper
 
 
-def decode_wrapper(cipher_fn):
+def decode_wrapper(cipher_fn: CipherFunction) -> CipherFunction:
     """Return wrapper for cipher functions to decode the output of the cipher function."""
 
-    def cipher_fn_wrapper(key, text, *args, **kwargs):
+    def cipher_fn_wrapper(key: int, text: int, *args: Any, **kwargs: Any) -> str:
         return decode(cipher_fn(key, text, *args, **kwargs))
 
     return cipher_fn_wrapper

--- a/src/util/min_bits.py
+++ b/src/util/min_bits.py
@@ -3,7 +3,7 @@
 from math import ceil, log2
 
 
-def min_bits(x):
+def min_bits(x: int) -> int:
     """Return the amount of bits needed to represent the given number.
 
     For example. to represent 8, 4 bits are needed: 0b1000.

--- a/src/util/rot.py
+++ b/src/util/rot.py
@@ -1,7 +1,7 @@
 """Exports functions which implement bit rotation."""
 
 
-def rot_left(bits, i, max_bit):
+def rot_left(bits: int, i: int, max_bit: int) -> int:
     """Bit rotation / Circular shift to the left.
 
     Shifts i bits to the left and adds the bits that fall of to the right side.
@@ -13,7 +13,7 @@ def rot_left(bits, i, max_bit):
     return ((bits % 2 ** (max_bit - i)) << i) | (bits >> (max_bit - i))
 
 
-def rot_right(bits, i, max_bit):
+def rot_right(bits: int, i: int, max_bit: int) -> int:
     """Bit rotation / Circular shift to the right.
 
     Shifts i bits to the right and adds the bits that fall of to the left side.

--- a/src/util/split.py
+++ b/src/util/split.py
@@ -1,7 +1,8 @@
 """Exports function which splits numbers into an array of bitstrings."""
+from typing import List
 
 
-def split(n, size, bits):
+def split(n: int, size: int, bits: int) -> List[int]:
     """Split a number into an array of bitstrings.
 
     Split the input bits into n bitstrings with given size. The most significant bits are returned first.

--- a/src/util/types.py
+++ b/src/util/types.py
@@ -1,6 +1,6 @@
 """Exports custom types used throughout the codebase."""
 
-from typing import Protocol, Any
+from typing import Protocol, Any, TypedDict, Callable
 
 
 class CipherFunction(Protocol):
@@ -9,3 +9,14 @@ class CipherFunction(Protocol):
     # But when hinting with Union[int, str] `mypy` also throws error when hinting a function with only int or str.
     # https://github.com/python/mypy/issues/7183
     def __call__(self, key: int, text: Any, *args: Any, **kwargs: Any) -> Any: ...
+
+
+Formatter = Callable[[Any], Any]
+
+CipherOptions = TypedDict('CipherOptions', {
+    '-o': str,
+    '--round-number': int,
+    '-m': str,
+    '-x': str,
+    'blocksize': int,
+})

--- a/src/util/types.py
+++ b/src/util/types.py
@@ -7,4 +7,5 @@ class CipherFunction(Protocol):
     # NOTE text should actually be hinted with Union[int,str]
     # NOTE return value should actually be hinted with Union[int,str]
     # But when hinting with Union[int, str] `mypy` also throws error when hinting a function with only int or str.
+    # https://github.com/python/mypy/issues/7183
     def __call__(self, key: int, text: Any, *args: Any, **kwargs: Any) -> Any: ...

--- a/src/util/types.py
+++ b/src/util/types.py
@@ -1,0 +1,10 @@
+"""Exports custom types used throughout the codebase."""
+
+from typing import Protocol, Any
+
+
+class CipherFunction(Protocol):
+    # TODO text should actually be hinted with Union[int,str]
+    # TODO return value should actually be hinted with Union[int,str]
+    # But when hinting with Union[int, str] `mypy` also throws error when hinting a function with only int or str.
+    def __call__(self, key: int, text: Any, *args: Any, **kwargs: Any) -> Any: ...

--- a/src/util/types.py
+++ b/src/util/types.py
@@ -4,7 +4,7 @@ from typing import Protocol, Any
 
 
 class CipherFunction(Protocol):
-    # TODO text should actually be hinted with Union[int,str]
-    # TODO return value should actually be hinted with Union[int,str]
+    # NOTE text should actually be hinted with Union[int,str]
+    # NOTE return value should actually be hinted with Union[int,str]
     # But when hinting with Union[int, str] `mypy` also throws error when hinting a function with only int or str.
     def __call__(self, key: int, text: Any, *args: Any, **kwargs: Any) -> Any: ...

--- a/src/util/wrap.py
+++ b/src/util/wrap.py
@@ -11,12 +11,15 @@ The only difference is that that now one can use larger strings since `ecb` will
 and pass them individually to `encrypt` and then return the concatentation of the encrypted blocks
 - which is exactly what ECB mode does.
 """
+from typing import Tuple, Any, Mapping
 
 from ciphers.modi.ecb import ecb
 from util.encode import encode_wrapper, decode_wrapper
+from util.types import CipherFunction, CipherOptions, Formatter
 
 
-def get_wrapped_cipher_functions(encrypt, decrypt, args):
+def get_wrapped_cipher_functions(encrypt: CipherFunction, decrypt: CipherFunction,
+                                 args: CipherOptions) -> Tuple[CipherFunction, CipherFunction]:
     """Return the correctly wrapped encryption and decryption functions."""
     """When parsing arguments, the following execution order has to be ensured:
         ===========================================================================
@@ -87,9 +90,9 @@ def get_wrapped_cipher_functions(encrypt, decrypt, args):
         """
 
     n = int(args['--round-number'])
-    ecb_mode = args['-m'] == 'ecb'
-    utf8_mode = args['-x'] == 'utf8'
-    blocksize = args['blocksize']
+    ecb_mode: bool = args['-m'] == 'ecb'
+    utf8_mode: bool = args['-x'] == 'utf8'
+    blocksize: int = args['blocksize']
 
     # Check if enum arguments are valid
     if args['-x'] not in ['utf8', 'none']:
@@ -117,7 +120,7 @@ def get_wrapped_cipher_functions(encrypt, decrypt, args):
 
     # Only format the output if we are not decrypting and using encoding since encoding would format the output
     #   itself already
-    _format = {'bin': bin, 'oct': oct, 'dec': str, 'hex': hex}
+    _format: Mapping[str, Formatter] = {'bin': bin, 'oct': oct, 'dec': str, 'hex': hex}
     # This should not be able to cause an KeyError because we already checked that all enum arguments are valid
     formatter = _format[args['-o']]
     w_encrypt = format_output_wrapper(w_encrypt, formatter)
@@ -127,24 +130,24 @@ def get_wrapped_cipher_functions(encrypt, decrypt, args):
     return w_encrypt, w_decrypt
 
 
-def format_output_wrapper(cipher_fn, formatter):
+def format_output_wrapper(cipher_fn: CipherFunction, formatter: Formatter) -> CipherFunction:
     """Return wrapper for cipher functions to cast the output into the specified format."""
 
-    def cipher_fn_wrapper(key, text, *args, **kwargs):
+    def cipher_fn_wrapper(key: int, text: Any, *args: Any, **kwargs: Any) -> Any:
         return formatter(cipher_fn(key, text, *args, **kwargs))
 
     return cipher_fn_wrapper
 
 
-def text_int_wrapper(cipher_fn):
+def text_int_wrapper(cipher_fn: CipherFunction) -> CipherFunction:
     """Return wrapper for cipher functions which casts text input to numbers."""
     return format_input_wrapper(cipher_fn, lambda text: int(text, 0))
 
 
-def format_input_wrapper(cipher_fn, formatter):
+def format_input_wrapper(cipher_fn: CipherFunction, formatter: Formatter) -> CipherFunction:
     """Return wrapper for cipher functions to cast cipher function text input into the specified format."""
 
-    def cipher_fn_wrapper(key, text, *args, **kwargs):
+    def cipher_fn_wrapper(key: int, text: int, *args: Any, **kwargs: Any) -> Any:
         return cipher_fn(key, formatter(text), *args, **kwargs)
 
     return cipher_fn_wrapper

